### PR TITLE
Add `CONTRIBUTING` section on PRs and git

### PR DESCRIPTION
**Pull request by sholderbach**
# Description

I describe the status quo (e.g. squashing) which came up as a FAQ.
Further more I add some opinionated best practices.

Feel free to correct me if I am overzealous or imprecise. As it is
already quite long I don't want to expand the text if possible.

# Developer-Facing Changes

Added devdocs to have more guidelines/FAQs addressed

### CODE REVIEW

**Review by amtoine — COMMENTED** [review]
only a bunch of really minor things i came up with, it's really great in the end :heart: 

one general thing about `CONTRIBUTING.md`: as it's becoming a bit larger now, maybe a simple table of content at the top would be great to jump to the section of interest :yum: 

> **Note**
> thanks again, i really find that super clear and great :star:

**Comment by fdncred** [comment]
+1 Thanks @sholderbach

**Comment by sholderbach** [comment]
Updated it locally based on your suggestions @amtoine. Will push later when I get back to stable WiFi.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -99,3 +99,86 @@ The most comprehensive test suite we have is the `nu-test-support` crate. For te
...
+
+### A good PR makes a change!
+
+As a result of this PR-centric strategy and the general goal that the reviewers should easily understand your change the PR title and description matters a great deal!
+
+Make sure your description is concise but contains all relevant information and context.
+This means demonstrating what changes, ideally through nushell code or output examples.
+Furthermore technical documentation you used or instructions for folks that want to play around with your change can help make the review process much easier.
+
+**Tip:** Try to follow the suggestions in our PR message template to make sure we can quickly focus on the technical merits or impact on the users.
```
[the remark this answers is not in the document]
Comment (sholderbach) [reply CONTRIBUTING.md:127]: You are the maestro of github flavor markdown.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -99,3 +99,86 @@ The most comprehensive test suite we have is the `nu-test-support` crate. For te
...
+
+#### A PR should limit itself to a single functional change or related set of same changes.
+
+Mixing different changes in the same PR will make the review process much harder and the PR might get stuck on one aspect while we would actually like to see another change land. Furthermore in case we are forced to revert a change, mixing and matching different aspects will make fixing bugs or regressions much harder.
+
+Thus, please try to **separate out unrelated changes**!
+**Don't** mix unrelated refactors with a potentially contested change.
+Stylistic fixes and housekeeping can be bundled up into singular PRs.
+
+#### Guidelines for the PR title
```
[the remark this answers is not in the document]
Comment (sholderbach) [reply CONTRIBUTING.md:137]: I don't think so we can easily have something productive there as the humans need to understand it in the end. 
I am personally in the camp of "traditional" git commit message.
So I find the "conventional commits" style of "refactor: refactor..." to add more noise than signal in a lot of cases. 

We were discussing having the work area in front of the colon as practiced by the serenity project but no consensus yet. 
 
https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -99,3 +99,86 @@ The most comprehensive test suite we have is the `nu-test-support` crate. For te
   cargo run --release -- --log-level trace --log-target file
   open $"($nu.temp-path)/nu-($nu.pid).log"
   ```
+
+## Git etiquette
+
+As nushell thrives on its broad base of volunteer contributors and maintainers with different backgrounds we have a few guidelines for how we best utilize git and GitHub for our contributions. We strive to balance three goals with those recommendations:
+
+1. The volunteer maintainers and contributors can easily follow what changes you are proposing with what impact and come to help you or make a decision if a change doesn't fit.
```
[the remark this answers is not in the document]
Comment (amtoine) [reply CONTRIBUTING.md:107]: yep, that's my concern here :wink:

File: CONTRIBUTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (amtoine) [reply CONTRIBUTING.md:127]: ehe, i discovered them not so long ago... i use them everywhere now :laughing: 
there really great to read imo

**Comment by amtoine** [comment]
> Updated it locally based on your suggestions @amtoine. Will push later when I get back to stable WiFi.
>
looking forward to seing the little tweaks :yum:

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -99,3 +99,86 @@ The most comprehensive test suite we have is the `nu-test-support` crate. For te
...
+  - We don't shame fixup commits while you try to figure out a problem. They can help others to see what you tried and didn't work. (see our squash policy)
+  - While active review is going on constant force pushing just to amend changes might be confusing
+    - GitHub's UI presents reviewers with less options to compare diffs
+    - fetched branches for experimentation become invalid
+    - the notification a maintainer receives has a low signal-to-noise
+  - Git pros can use their judgement to rebase/squash to clean up the history *if it aids the understanding* of a larger change during review 
+- Merge conflicts:
+  - In general you should take care of resolving merge conflicts if possible.
+    - You can use your judgement whether to `git merge main` or to `git rebase main`
+    - Choose what simplifies having confidence in the conflict resolution and the review.
```
[the remark this answers is not in the document]
Comment (sholderbach) [reply CONTRIBUTING.md:182]: Your making absolutely valid arguments for a no public rebase policy, but I think that is a bridge too far as we also probably have contributors that come from rebase heavy workflows.

I added a sentence that we officially bless merge commits in feature branches.

But there is a place for crafting nicer histories with rebase. Furthermore I like an interactive rebase when facing a semantic merge conflict to see if each commit still works with the upstream changes.

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (amtoine) [reply CONTRIBUTING.md:137]: > I don't think so we can easily have something productive there as the humans need to understand it in the end.
>
yeah of course...

> I am personally in the camp of "traditional" git commit message. 
>
I am personally in the camp of meaningfull but still complete `git` commit message :wink: 
which is not something easy to master at all!!

> So I find the "conventional commits" style of "refactor: refactor..." to add more noise than signal in a lot of cases.
> 
yeah `refactor: refactor ...` is not great... :sweat_smile: 
i tend to write, `refactor: move files around` or `refactor: simplify that function` or `refactor: compress duplicate logic`, ..., that kind of messages :thinking: 

> We were discussing having the work area in front of the colon as practiced by the serenity project but no consensus yet.
> 
> https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md
>
i kinda like their `:` notation :open_mouth:

File: CONTRIBUTING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (amtoine) [reply CONTRIBUTING.md:182]: > Your making absolutely valid arguments for a no public rebase policy, but I think that is a bridge too far as we also probably have contributors that come from rebase heavy workflows.
> 
yeah for sure, already seen that and it's 100% fine.
people need to come and contribute, regardless their `git` background and merge strategies!

> I added a sentence that we officially bless merge commits in feature branches.
> 
that is a very good compromise :ok_hand: 

> But there is a place for crafting nicer histories with rebase.
>
i would say this is the work of the PR merge strategy set to a squash rebase with no merge commit, right?
this is how i see it for now at least :relieved:  

> Furthermore I like an interactive rebase when facing a semantic merge conflict to see if each commit still works with the upstream changes.
>
ohh yeah that's a good point :+1: 

but i can not forget the fact that i find rebasing a PR feature branch to be a pain to deal with as a reviewer :sob:

**Pull request merged**